integration-tests/src/tests/client/features/congestion_control.rs

```diff
@@ -0,0 +1,228 @@
+        chunk_header
+            .congestion_info()
+            .expect("chunk header must have congestion info after upgrade");
+    }
+    let tip = env.clients[0].chain.head().unwrap();
+
+    // check sender shard is still sending transactions from the pool
+    let sender_shard_id =
+        env.clients[0].epoch_manager.account_id_to_shard_id(&sender_id, &tip.epoch_id).unwrap();
+    let sender_shard_chunk_header =
+        &chunks.get(sender_shard_id as usize).expect("chunk must be available");
+    let sender_chunk =
+        env.clients[0].chain.get_chunk(&sender_shard_chunk_header.chunk_hash()).unwrap();
+    assert!(
+        sender_chunk.transactions().len() > 0,
+        "test setup error: sender is out of transactions"
+    );
+    assert!(
+        sender_chunk.prev_outgoing_receipts().len() > 0,
+        "test setup error: sender is not sending receipts"
+    );
+
+    // check congestion is observed on the contract's shard
+    let contract_shard_id =
+        env.clients[0].epoch_manager.account_id_to_shard_id(&contract_id, &tip.epoch_id).unwrap();
+    assert!(
+        contract_shard_id != sender_shard_id,
+        "test setup error: sender and contract are on the same shard"
+    );
+    let contract_shard_chunk_header =
+        &chunks.get(contract_shard_id as usize).expect("chunk must be available");
+    let _congestion_info = contract_shard_chunk_header.congestion_info().unwrap();
+    // TODO(congestion_control) - properly initialize
+    // assert_eq!(
+    //     congestion_info.congestion_level(),
+    //     1.0,
+    //     "contract's shard should be fully congested"
+    // );
+}
+
+/// Check we are still in the old version and no congestion info is shared.
+#[track_caller]
+fn check_old_protocol(env: &TestEnv) {
+    assert!(
+        !ProtocolFeature::CongestionControl.enabled(env.get_head_protocol_version()),
+        "test setup error: chain already updated to new protocol"
+    );
+    let block = env.clients[0].chain.get_head_block().unwrap();
+    let chunks = block.chunks();
+    assert!(chunks.len() > 0, "no chunks in block");
+    for chunk_header in chunks.iter() {
+        assert!(
+            chunk_header.congestion_info().is_none(),
+            "old protocol should not have congestion info but found {:?}",
+            chunk_header.congestion_info()
+        );
+    }
+}
+
+/// Create a function call that has 100 Tgas attached and will burn it all.
+fn new_fn_call_100tgas(
+    nonce_source: &mut u64,
+    signer: &InMemorySigner,
+    contract_id: AccountId,
+    block_hash: near_primitives::hash::CryptoHash,
+) -> SignedTransaction {
+    let hundred_tgas = 100 * 10u64.pow(12);
+    let deposit = 0;
+    let nonce = *nonce_source;
+    *nonce_source += 1;
+    SignedTransaction::call(
+        nonce,
+        signer.account_id.clone(),
+        contract_id,
+        signer,
+        deposit,
+        // easy way to burn all attached gas
+        "loop_forever".to_owned(),
```

Comment:
We could use a host function "burn_gas" that would burn gas but not waste the time and CPU to do so ;) Absolutely no need to do it in this PR thought.

Reply:
yeah, kind of hurts to let the CPU spin round and round ^^

Good idea to add a host function! Something to maybe discuss with the runtime team how to do it best, adding it to the runtime interface with a NEP and protocol upgrade seems a bit overkill, so maybe it could be a test_features-only thing.